File: 797-rabbits-in-forest/rabbits-in-forest.cpp
class Solution {
public:

    // Solution
    int numOfRabbits(const std::vector<int>& rabbits) {
        int totalRabbits = 0;
        std::unordered_map<int,int>countMap;

        for(int rabbit: rabbits) {
            countMap[rabbit]++;
        }

        for(const auto &entry:countMap) {
            int reportedOthers = entry.first;
            int reportCount = entry.second;
            int groupSize = reportedOthers + 1;
            int groupNeeded = std::ceil((double)reportCount / groupSize);

            totalRabbits += groupSize * groupNeeded;
        }

        return totalRabbits;
    }
    
    /*
        Analysis:
        Time Complexity  : O(N)
        Space Complexity : O(N)
    */

    int numRabbits(vector<int>& answers) {
        return numOfRabbits(answers);
    }
};

File: 797-rabbits-in-forest/rabbits-in-forest.java
class Solution {

    // Function that calculates the minimum number of rabbits in forest
    public int numOfRabbits(int[] rabbits) {
        int totalRabbits = 0;
        HashMap<Integer, Integer> countMap = new HashMap<Integer, Integer>();

        for (int rabbit:rabbits) {
            countMap.put(rabbit, countMap.getOrDefault(rabbit,0)+1);
        }

        for(int key:countMap.keySet()) {
            int groupSize = key + 1;
            int groupNeeded = (int)Math.ceil(countMap.get(key)*1.0 / groupSize * 1.0);
            totalRabbits += groupSize * groupNeeded;
        }
        return totalRabbits;
    }

    /*
        Analysis:
        Time Complexity  : O(N)
        Space Complexity : O(N)
    */

    public int numRabbits(int[] answers) {
        return numOfRabbits(answers);
    }
}